optimism: bump gaslimit to 10 million #870

tynes opened the pull request:
<!--
Please fill in each sections of this template, and delete any sections that are not relevant.

Need help?
Refer to our contributing guidelines for additional information about making a good pull request:
https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**
This PR bumps the protocol level `gasLimit` to 10 million. This is the maximum amount of gas that a single transaction can consume. This number shouldn't grow too large relative to the amount of gas consumed by a block. If the gas consumed by a single transaction on Optimism consumes more gas than what can be fit into an L1 block, then it will not be possible to execute the fraud proof based on the current construction.

**Additional context**
I believe that these are the only 2 places required to be changed. There is also a `gasLimit` in the CTC for `enqueue` transactions that is configured separately and an open PR for removing the need to configure the `gasLimit` as part of the state dump process. See https://github.com/ethereum-optimism/optimism/pull/840

This PR along with Cached Safety Checking will make contract deployment much better.
For historical info on cached safety checking, see:
- https://github.com/ethereum-optimism/contracts/pull/310
- https://github.com/ethereum-optimism/contracts/pull/253

smartcontracts commented:
Following lines should be updated as well:
* https://github.com/ethereum-optimism/optimism/blob/0c1680533c870efed3e5335f53c03996413daf4e/packages/contracts/tasks/deploy.ts#L8
* https://github.com/ethereum-optimism/optimism/blob/0c1680533c870efed3e5335f53c03996413daf4e/packages/contracts/tasks/deploy.ts#L10

Following lines *might* need to be updated:
* https://github.com/ethereum-optimism/optimism/blob/0c1680533c870efed3e5335f53c03996413daf4e/l2geth/scripts/start.sh#L9
* https://github.com/ethereum-optimism/optimism/blob/develop/specs/l2geth/sequencer/fees.md#L63

tynes commented:
cc @smartcontracts  https://github.com/ethereum-optimism/optimism/blob/8d67991aba584c1703692ea46273ea8a1ef45f56/packages/data-transport-layer/src/utils/constants.ts#L1

tynes commented:
Is there any way that we can deduplicate some of these?

elenadimitrova commented on packages/contracts/src/contract-dumps.ts:
```diff
@@ -115,7 +115,7 @@ export const makeStateDump = async (cfg: RollupDeployConfig): Promise<any> => {
     ovmGasMeteringConfig: {
       minTransactionGasLimit: 0,
-      maxTransactionGasLimit: 9_000_000,
+      maxTransactionGasLimit: 10_000_000,
```
Since we've upped the `maxTransactionGasLimit` in EM we might want to up the `L2_GAS_DISCOUNT_DIVISOR` as well as now the burned gas will be higher for txs sent with a default gas limit. Or at least we could make a note to consider this as part of the work in #762

tynes replied:
Is it ok if we handle that in #762?

elenadimitrova commented on ops/envs/geth.env:
```diff
@@ -29,5 +29,5 @@ IPC_DISABLE=true
 NO_USB=true
 NO_DISCOVER=true
-TARGET_GAS_LIMIT=9000000
+TARGET_GAS_LIMIT=10000000
```
Just a question on how the update to `ExecutionManager.maxTransactionGasLimit` was done [here](https://github.com/ethereum-optimism/optimism/pull/840/files#diff-17b617e193a2e9627cf0c5459c05a94f7de9ce2ebde734ede16251f1a7a16118R329) since I am reading `maxTransactionGasLimit` being stored in slot `2` not `4`. What am I missing?

tynes replied:
We spoke out of band and have agreed that the slot is `4`, see https://github.com/ethereum-optimism/optimism/pull/840

tynes commented:
I've rebased and updated the instances of 9 million to 10 million.

> Is there any way that we can deduplicate some of these?

Leaving the above comment for future work, but we should really be able to have a single source of truth that is used throughout the rest of the system

platocrat commented:
@tynes what's preventing this from being merged?

maurelian commented:
I don't see it mentioned, but assume that this change is due to the recent L1 block gas limit increasing to [15 million](https://etherscan.io/chart/gaslimit). 
Do we have a rule of thumb or other means for determining the safe L2 gas limit?

tynes commented:
> I don't see it mentioned, but assume that this change is due to the recent L1 block gas limit increasing to [15 million](https://etherscan.io/chart/gaslimit).
> Do we have a rule of thumb or other means for determining the safe L2 gas limit?

This change is due to the increase in L1 block gas limit as well as the large gas overhead that causes projects to have to refactor their code to be able to deploy on Optimism. The larger that our gas limit is, the more risky it becomes when executing a fraud proof because it will require a larger percentage of the total block's gas and it must fit within a single block.

We should be generally cautious around raising the gas limit and reserve the right to lower it in the future if we feel like we need to. A fix to the sequencer entrypoint where it introduces less overhead would let us safely decrease the gas limit.

This particular raise of the gas limit allows uniswap v3 to be deployed without refactoring of the contracts.

snario commented:
This needs to re-target `regenesis/0.4.0`.
